src/common/variant8.h

```diff
@@ -13,54 +13,213 @@
 #define VARIANT8_ERROR 0xff // error
 
+//pointer types
+#define VARIANT8_PI8   (VARIANT8_I8   | VARIANT8_PTR)
+#define VARIANT8_PUI8  (VARIANT8_UI8  | VARIANT8_PTR)
+#define VARIANT8_PI16  (VARIANT8_I16  | VARIANT8_PTR)
+#define VARIANT8_PUI16 (VARIANT8_UI16 | VARIANT8_PTR)
+#define VARIANT8_PI32  (VARIANT8_I32  | VARIANT8_PTR)
+#define VARIANT8_PUI32 (VARIANT8_UI32 | VARIANT8_PTR)
+#define VARIANT8_PFLT  (VARIANT8_FLT  | VARIANT8_PTR)
+#define VARIANT8_PCHAR (VARIANT8_CHAR | VARIANT8_PTR)
+
+//variant errors
+#define VARIANT8_ERR_MALLOC   1 // memory allocation error (during conversion to strings and allocating pointer types)
+#define VARIANT8_ERR_UNSTYP   2 // unsupported conversion (during conversion)
+#define VARIANT8_ERR_UNSCON   3 // unsupported conversion (during conversion)
+#define VARIANT8_ERR_INVFMT   4 // invalid format (during conversion from string)
+#define VARIANT8_ERR_OOFRNG   5 // out of range (during conversion from bigger to lower range number)
+
+//macros for variant8 structure constants
+#define _VARIANT8_TYPE(type, _8, _16, _32) ((variant8_t){ type, _8, { _16 }, { _32 } })
+#define _VARIANT8_EMPTY() _VARIANT8_TYPE(VARIANT8_EMPTY, 0, 0, 0)
+
 #pragma pack(push)
 #pragma pack(1)
 
 typedef struct _variant8_t {
     uint8_t type;
     uint8_t usr8;
-    uint16_t usr16;
     union {
-        char *pc;
+        uint16_t usr16;
+        uint16_t size;
+        uint16_t err16;
+    };
+    union {
+        void *ptr;
+        char *pch;
+        float *pflt;
+        uint32_t *pui32;
+        int32_t *pi32;
+        uint16_t *pui16;
+        int16_t *pi16;
+        uint8_t *pui8;
+        int8_t *pi8;
         uint32_t usr32;
+        char ch;
         float flt;
         uint32_t ui32;
         int32_t i32;
         uint16_t ui16;
         int16_t i16;
         uint8_t ui8;
         int8_t i8;
+        uint32_t err32;
     };
 } variant8_t;
 
 #pragma pack(pop)
 
+
 #ifdef __cplusplus
+
+class cvariant8 : public variant8_t {
+public: // construction/destruction
+    cvariant8();
+    cvariant8(const cvariant8& var8);
+    cvariant8(int8_t val);
+    cvariant8(uint8_t val);
+    cvariant8(int16_t val);
+    cvariant8(uint16_t val);
+    cvariant8(int32_t val);
+    cvariant8(uint32_t val);
+    cvariant8(float val);
+    cvariant8(const char* val);
+    ~cvariant8();
+public: // public functions
+    cvariant8 copy();
+    cvariant8& attach(variant8_t var8);
+    variant8_t detach();
+    cvariant8& change_type(uint8_t new_type);
+public: // assignment operators
+    cvariant8& operator = (const cvariant8& var8);
+    cvariant8& operator = (int8_t val);
+    cvariant8& operator = (uint8_t val);
+    cvariant8& operator = (int16_t val);
+    cvariant8& operator = (uint16_t val);
+    cvariant8& operator = (int32_t val);
+    cvariant8& operator = (uint32_t val);
+    cvariant8& operator = (float val);
+    cvariant8& operator = (const char* val);
+public: // extractors
+    operator int8_t() { return (type == VARIANT8_I8)?i8:0; }
```

Comment:
I'd recommend using a member function like
inline uint8_t AsUint8t(uint8_t default_value)const { return (type == VARIANT8_U8)?u8:default_value; }
benefits:
- it makes clear, that some default value is being returned in case of an error
- it may be easier to use in cases, where multiple overloaded operators might be applicable in a particular context - cases, that may be solved by using static_cast<> which doesn't look too nice